pkg/controller/egress/controller.go

```diff
@@ -449,3 +456,61 @@ func (c *EgressController) enqueueEgresses(poolName string) {
 		c.queue.Add(egress.Name)
 	}
 }
+
+func (c *EgressController) updateEgressAllocatedCondition(egress *egressv1beta1.Egress, err error) {
+	var desiredCondition *egressv1beta1.EgressCondition
+	if egress.Spec.ExternalIPPool != "" {
+		if err == nil {
+			desiredCondition = &egressv1beta1.EgressCondition{
+				Type:               egressv1beta1.IPAllocated,
+				Status:             v1.ConditionTrue,
+				Reason:             "Allocated",
+				Message:            "EgressIP is successfully allocated",
+				LastTransitionTime: metav1.Now(),
+			}
+		} else {
+			desiredCondition = &egressv1beta1.EgressCondition{
+				Type:               egressv1beta1.IPAllocated,
+				Status:             v1.ConditionFalse,
+				Reason:             "AllocationError",
+				Message:            fmt.Sprintf("Cannot allocate EgressIP from ExternalIPPool %s due to: %v", egress.Spec.ExternalIPPool, err),
+				LastTransitionTime: metav1.Now(),
+			}
+		}
+	}
+
+	toUpdate := egress.DeepCopy()
+	var updateErr, getErr error
+	if err := retry.RetryOnConflict(retry.DefaultRetry, func() error {
+		actualCondition, index := getIPAllocatedCondition(toUpdate)
+		if actualCondition != nil && actualCondition.Status == desiredCondition.Status && actualCondition.Reason == desiredCondition.Reason && actualCondition.Message == desiredCondition.Message {
+			return nil
+		}
+		if desiredCondition == nil {
+			return nil
+		}
+		if index != -1 {
+			toUpdate.Status.Conditions[index] = *desiredCondition
+		} else {
+			toUpdate.Status.Conditions = append(toUpdate.Status.Conditions, *desiredCondition)
+		}
```

Comment:
When `desiredCondition` is `nil` but `actualCondition` is not `nil`, we need to delete the condition. A simpler to handle three cases is:
```
var newConditions []egressv1beta1.EgressCondition
for _, c := range toUpdate.Status.Conditions {
        if c.Type != egressv1beta1.IPAllocated {
                newConditions = append(newConditions, c)
        }
}
if desiredCondition != nil {
        newConditions = append(newConditions, desiredCondition)
}
```